Login: Update Networking and Yosemite layers to handle Jetpack connection data API

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #7597 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR continues the work in #7600 to integrate the [Jetpack connection data API](https://github.com/Automattic/jetpack/blob/trunk/docs/rest-api.md#get-wp-jsonjetpackv4connectiondata). 

This API also requires cookie authentication, so I've also updated the `DeauthenticatedStore`, `DeauthenticatedState`, `StoresManager` and `DefaultStoresManager` to make it possible to inject `Network` into the store and make it testable.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
The changes haven't been integrated into the UI so just CI passing is sufficient.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Add new models for Jetpack connection data

itsmeichigo committed:
Add mapper for JetpackConnectedUser

itsmeichigo committed:
Add test response for JetpackConnectedUser

itsmeichigo committed:
Add tests for JetpackConnectedUserMapper

itsmeichigo committed:
Rename JetpackConnectedUser to JetpackUser

itsmeichigo committed:
Add new endpoint for fetching Jetpack user

itsmeichigo committed:
Update tests for JetpackConnectionRemote

itsmeichigo committed:
Add new action and store for fetching Jetpack user

itsmeichigo committed:
Update deauthenticated store and state to make deauthenticated store testable

itsmeichigo committed:
Add tests for JetpackConnectionStore

itsmeichigo committed:
Update typo in comment

jaclync commented on Networking/Networking/Model/WordPressComUser.swift:
```diff
@@ -0,0 +1,65 @@
+import Foundation
+
+/// Basic information of a WordPress.com user
+public struct WordPressComUser: Decodable, GeneratedFakeable, GeneratedCopiable {
```
nit: since we already have `DotcomDevice`, does it make sense to keep the naming consistent?
```suggestion
public struct DotcomUser: Decodable, GeneratedFakeable, GeneratedCopiable {
```

jaclync commented on Networking/Networking/Model/WordPressComUser.swift:
```diff
@@ -0,0 +1,65 @@
+        textDirection = try container.decode(String.self, forKey: .textDirection)
+        siteCount = try container.decode(Int64.self, forKey: .siteCount)
+        avatar = try? container.decode(String.self, forKey: .avatar)
+    }
```
nit: it seems like there's no custom logic decoding all the fields in the response, maybe we don't need this implementation and it might just work since it conforms to `Decodable`?

jaclync commented on Networking/Networking/Model/WordPressComUser.swift:
```diff
@@ -0,0 +1,65 @@
+    public let textDirection: String
+
+    /// Number of registered sites in WP.com
+    public let siteCount: Int64
```
nit: I don't know how this struct is going to be used in the future, just wondering if we plan to use all the properties

jaclync commented on Networking/Networking/Model/JetpackUser.swift:
```diff
@@ -0,0 +1,53 @@
+    public let isConnected: Bool
+
+    /// Whether the user is the primary account connected to the site's Jetpack
+    public let isMaster: Bool
```
super nit: might be more friendly and easier to understand (or like `isPrimaryUser`/`isPrimaryAccount`)

```suggestion
    public let isPrimary: Bool
```

jaclync commented on Networking/Networking/Model/JetpackUser.swift:
```diff
@@ -0,0 +1,53 @@
+        username = try container.decode(String.self, forKey: .username)
+        wpcomUser = try? container.decode(WordPressComUser.self, forKey: .wpcomUser)
+        gravatar = try? container.decode(String.self, forKey: .gravatar)
+    }
```
similar to the other comment, maybe this method isn't necessary because there's no custom decoding logic and all properties are decodable

jaclync commented on Networking/Networking/Remote/JetpackConnectionRemote.swift:
```diff
@@ -19,16 +19,26 @@ public final class JetpackConnectionRemote: Remote {
 private extension JetpackConnectionRemote {
     enum Path {
         static let jetpackConnectionURL = "/jetpack/v4/connection/url"
+        static let jetpackConnectionData = "/jetpack/v4/connection/data"
```
super nit: how about mentioning it's to fetch the Jetpack user given its usage in `fetchJetpackUser`?
```suggestion
        static let jetpackConnectionUser = "/jetpack/v4/connection/data"
```

jaclync commented on Networking/Networking/Remote/JetpackConnectionRemote.swift:
```diff
@@ -19,16 +19,26 @@ public final class JetpackConnectionRemote: Remote {
 public extension JetpackConnectionRemote {
     enum ConnectionError: Int, Error {
         case malformedURL
+        case currentUserNotFound
```
nit: is this error case used somewhere?

jaclync commented on Networking/NetworkingTests/Remote/JetpackConnectionRemoteTests.swift:
```diff
@@ -53,4 +53,45 @@ final class JetpackConnectionRemoteTests: XCTestCase {
         XCTAssertEqual(result.failure as? NetworkError, error)
     }
+
+    func test_fetchJetpackUser_correctly_returns_parsed_url() throws {
```
nit: just wondering about the name, it looks like it's returning the user instead of a url?

jaclync commented on Yosemite/Yosemite/Base/StoresManager.swift:
```diff
@@ -27,6 +28,10 @@ public protocol StoresManager {
 
+    /// If the state is deauthenticated, send it a siteURL and authenticator.
+    ///
+    func updateDeauthenticatedState(with siteURL: String, authenticator: Authenticator)
```
I'm a bit lost here and since this isn't used so far, why would a site URL be relevant in the deauthenticated state? 🤔 this also feels like an implementation detail, does this have to be in the protocol? again, I don't know how it's called later and some more context would be helpful.

itsmeichigo replied:
Sorry for not explaining clearly earlier. The idea is to send the site URL and authenticator before making any request to the WP.org REST API. The URL and authenticator are needed for handling cookie authentication. 

I've updated the comment on `updateDeauthenticatedState` to state that. Let me know if it doesn't make sense.

jaclync replied:
thanks for the context! responded in another related thread https://github.com/woocommerce/woocommerce-ios/pull/7663#discussion_r964377343

itsmeichigo committed:
Generate copiable for JetpackUser and WordPressComUser

jaclync commented on WooCommerce/Classes/Yosemite/DeauthenticatedState.swift:
```diff
@@ -12,12 +11,20 @@ class DeauthenticatedState: StoresManagerState {
+    ///
+    func updateStores(with siteURL: String, network: Network) {
+        for store in services {
+            store.updateRemote(with: siteURL, network: network)
```
would any of the `DeauthenticatedStore`s work without the siteURL and network? if not, how about making `services: [DeauthenticatedStore]` a variable and then `remote: JetpackConnectionRemote` become a non-nil constant in `JetpackConnectionStore` and `DeauthenticatedStore.updateRemote` isn't necessary anymore?

itsmeichigo replied:
Even though siteURL and network are required for `JetpackConnectionStore`, I can't come up with a good solution to keep them in the store's initializer. We don't have any site URL and network immediately when initializing `DeauthenticatedState`,  only immediately when there's a need to work with WP.org API. That's why I decided to add this method to update the store later.

Please share your thought if you have a better solution for this.

jaclync replied:
do we expect more than just `JetpackConnectionStore` that requires the same `siteURL` and `network` dependencies at a later time? if not, how about creating an action like `JetpackConnectionAction.update(siteURL: String, network: Network)` that initializes the `JetpackConnectionRemote` internally so that the async update is just an implementation detail within `JetpackConnectionStore`?

itsmeichigo replied:
Good idea! We may need to use WP.org API to install/activate Jetpack the plugin later but I suppose we can add those into this same `JetpackConnectionStore`. My only concern is that we'll need to initialize `WordPressOrgNetwork` from the UI layer when calling this action, but we can just import the class for that so I guess it's fine. Updated this in the later commits.

jaclync replied:
> My only concern is that we'll need to initialize WordPressOrgNetwork from the UI layer when calling this action, but we can just import the class for that so I guess it's fine. Updated this in the later commits.

I'm not sure how the update action is used later, does the `Network` have to be initialized in the app/UI layer? Does it change over time together with `siteURL`? We can always see how we can improve the setup later when the actions are integrated with UI.

jaclync commented on Yosemite/Yosemite/Actions/JetpackConnectionAction.swift:
```diff
@@ -1,8 +1,11 @@
     /// Fetches the URL used for setting up Jetpack connection using the given authenticator
-    case fetchJetpackConnectionURL(siteURL: String, authenticator: Authenticator, completion: (Result<URL, Error>) -> Void)
+    case fetchJetpackConnectionURL(completion: (Result<URL, Error>) -> Void)
+    /// Fetches the user connection state with the given site's Jetpack and authenticator.
```
super nit: does this action still need the authenticator as mentioned in the comment?

jaclync reviewed (commented):
Took a pass and had some clarifying questions about how the update with site URL works to get more context about the changes. Please ping me on the PR ASAP so I'll take another pass and help wrap up the PR before I go AFK tomorrow!

itsmeichigo committed:
Rename WordPressComUser to DotComUser

itsmeichigo committed:
Use default decoding logic

itsmeichigo committed:
Remove unused properties for DotComUser

itsmeichigo committed:
Rename jetpackConnectionUser endpoint and remove unused error

itsmeichigo committed:
Rename test to match the tested subject

itsmeichigo committed:
Update comment explaining the purpose of updateDeauthenticatedState

itsmeichigo committed:
Update comments on JetpackConnectionAction

itsmeichigo committed:
Update JetpackUserMapperTests

itsmeichigo commented:
@jaclync Thank you for the review! I've addressed most of the comments above in the subsequent commits.

Regarding the injection of site URL and authenticator - I haven't come up with a more decent solution since we don't have those when initializing the `JetpackConnectionStore`. Let me know what you think.

jaclync commented on Networking/Networking/Model/Copiable/Models+Copiable.generated.swift:
```diff
@@ -169,6 +169,30 @@ extension CouponReport {
     }
 }
 
+extension DotComUser {
```
super nit: I think the existing `Dotcom*` structs have `com` as lowercase, super minor though

jaclync commented:
@itsmeichigo responded to your comments, and resolved most of the comments 🙂

itsmeichigo committed:
Rename DotComUser to DotcomUser

itsmeichigo committed:
Restore custom decoding initializer for JetpackUser to fix test failure

itsmeichigo committed:
Remove updateDeauthenticatedState

itsmeichigo committed:
Remove updateStores from DeauthenticatedState

itsmeichigo committed:
Add action to update remote for JetpackConnectionAction

itsmeichigo committed:
Update unit tests for JetpackConnectionStore

itsmeichigo committed:
Remove white line

itsmeichigo commented:
@jaclync I've switched to update `JetpackConnectionStore` through an action and revert previous changes as you suggested. This PR is ready for another look 🙇.

jaclync commented on Yosemite/YosemiteTests/Stores/JetpackConnectionStoreTests.swift:
```diff
@@ -0,0 +1,118 @@
+        XCTAssertEqual(result.failure as? NetworkError, error)
+    }
+
+    func test_fetchJetpackUser_correctly_returns_parsed_url() throws {
```
super nit:
```suggestion
    func test_fetchJetpackUser_correctly_returns_parsed_user() throws {
```

jaclync commented on Yosemite/Yosemite/Model/Mocks/MockStoresManager.swift:
```diff
@@ -1,5 +1,6 @@
 import Combine
 import Storage
+import WordPressKit
```
nit: this import might not be necessary anymore?

jaclync commented on Yosemite/Yosemite/Base/StoresManager.swift:
```diff
@@ -1,5 +1,6 @@
 import Combine
 import Foundation
+import WordPressKit
```
nit: this import might not be necessary anymore?

jaclync commented on WooCommerce/Classes/Yosemite/DefaultStoresManager.swift:
```diff
@@ -199,7 +201,7 @@ class DefaultStoresManager: StoresManager {
     /// Updates the user roles for the default Store site.
     ///
-    func updateDefaultRoles(_ roles: [User.Role]) {
+    func updateDefaultRoles(_ roles: [Yosemite.User.Role]) {
```
nit: are diffs in this file still necessary?

jaclync commented on WooCommerce/Classes/Yosemite/DeauthenticatedState.swift:
```diff
@@ -1,7 +1,6 @@
 import Yosemite
-
-
+import protocol Networking.Network
```
nit: this import might not be necessary anymore?

jaclync commented on Yosemite/Yosemite/Actions/JetpackConnectionAction.swift:
```diff
@@ -1,8 +1,14 @@
 import Foundation
+import Networking
```
super nit: maybe just `import protocol Networking.Network` is sufficient?

jaclync commented on Networking/Networking/Model/JetpackUser.swift:
```diff
@@ -0,0 +1,52 @@
+        self.gravatar = gravatar
+    }
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        isConnected = try container.decode(Bool.self, forKey: .isConnected)
+        isPrimary = try container.decode(Bool.self, forKey: .isPrimary)
+        username = try container.decode(String.self, forKey: .username)
+        wpcomUser = try? container.decode(DotcomUser.self, forKey: .wpcomUser)
+        gravatar = try? container.decode(String.self, forKey: .gravatar)
+    }
```
nit: this might not be necessary?

itsmeichigo replied:
I had to restore this since we need to make `DotcomUser` parsing failable with the optional `try?`. Otherwise, in cases where the user is not connected to Jetpack, parsing `JetpackUser` will fail too.

jaclync reviewed (approved):
Thanks for the quick updates @itsmeichigo! LGTM now :shipit:

itsmeichigo committed:
Rename test in JetpackConnectionStoreTests

itsmeichigo committed:
Add typealias for DotComUser and JetpackUser

itsmeichigo committed:
Remove redundant imports

itsmeichigo commented:
> I'm not sure how the update action is used later, does the Network have to be initialized in the app/UI layer? Does it change over time together with siteURL? We can always see how we can improve the setup later when the actions are integrated with UI.

@jaclync the `WordPressOrgNetwork` needs `Authenticator` to be initialized. We can inject the authenticator through the action, but this makes the `JetpackConnectionStore` not testable since the network will then be initialized based on the authenticator.

I'll keep this implementation and see if it needs to be updated when integrated with the UI layer.
